url73e3057d61_empeeoh_BACnet4J_tgz-pJ8-com_serotonin_bacnet4j_test_OverflowTestJ8/src/com/serotonin/bacnet4j/npdu/IncomingRequestParser.java

```diff
@@ -0,0 +1,95 @@
+
+abstract public class IncomingRequestParser {
+    private final Network network;
+    protected final ByteQueue originalQueue;
+    protected final ByteQueue queue;
+    protected Address from;
+    protected OctetString linkService;
+    protected ServicesSupported servicesSupported;
+
+    protected IncomingRequestParser(Network network, ByteQueue queue, OctetString localFrom) {
+        this.network = network;
+        this.queue = queue;
+        this.originalQueue = (ByteQueue) queue.clone();
+        linkService = localFrom;
+    }
+
+    public void run() {
+        try {
+            // Determine the supported services. Logical branch supports testing.
+            if (network == null || network.getTransport() == null) {
+                // Testing
+                servicesSupported = new ServicesSupported();
+                servicesSupported.setAll(true);
+            }
+            else
+                servicesSupported = network.getTransport().getLocalDevice().getServicesSupported();
+
+            parseFrame();
+
+            // Create the APDU.
+            APDU apdu = parseApdu();
+            if (apdu != null && network != null) {
+                if (network.isLocal(from))
+                    linkService = null;
+                network.getTransport().incomingApdu(apdu, from, linkService);
+            }
+        }
+        catch (Exception e) {
+            ExceptionDispatch.fireReceivedException(e);
+        }
+        catch (Throwable t) {
+            ExceptionDispatch.fireReceivedThrowable(t);
+        }
+    }
+
+    /**
+     * Parse out the network-specific data up until the NPCI.
+     * 
+     * @throws MessageValidationAssertionException
+     */
+    abstract protected void parseFrame() throws MessageValidationAssertionException;
+
+    public APDU parseApdu() throws Exception {
+        // Network layer protocol control information. See 6.2.2
+        NPCI npci = new NPCI(queue);
+        if (npci.getVersion() != 1)
+            throw new MessageValidationAssertionException("Invalid protocol version: " + npci.getVersion());
+        if (npci.isNetworkMessage())
+            return null; // throw new MessageValidationAssertionException("Network messages are not supported");
+
+        // Check the destination network number work and do not respond to foreign networks requests  
+        if (npci.hasDestinationInfo()) {
+            int destNet = npci.getDestinationNetwork();
+            if (destNet > 0 && destNet != 0xffff && network.getLocalNetworkNumber() > 0
+                    && network.getLocalNetworkNumber() != destNet)
+                return null;
+        }
+
+        if (npci.hasSourceInfo())
+            from = new Address(npci.getSourceNetwork(), npci.getSourceAddress());
+        else
+            from = new Address(linkService);
+
+        // Create the APDU.
+        try {
+            return APDU.createAPDU(servicesSupported, queue);
+        }
+        catch (BACnetException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)